File: fsharp/saddle-points/SaddlePoints.fs
module SaddlePoints

let saddlePoints (matrix : int list list) : (int * int) list =
    match matrix with 
    | [] | [[]] -> []
    | _ -> 
        let maxes = matrix |> List.map List.max

        let mins = 
            [ for j, _ in List.indexed matrix.Head do 
                yield [ for row in matrix do yield row.[j]]]
            |> List.map List.min

        [ for i, row in List.indexed matrix do 
            for j, el in List.indexed row do 
                if el = maxes.[i] && el = mins.[j] then yield i, j ]

File: ruby/saddle-points/saddle_points.rb
class Matrix
  attr_reader :rows, :columns

  def initialize(raw_matrix)
    @rows =
      raw_matrix
        .lines
        .map { |l| l.split.map(&:to_i) }

    @columns = @rows.transpose
  end

  def saddle_points
    results = []

    @rows.each_index do |x|
      @columns.each_index do |y|
        results << [x, y] if @rows[x].max == @columns[y].min
      end
    end

    results
  end
end
